File: Secure Code Review/SQL/.Net/sqli.cs
using System;
using System.Data.SQLite;

class VulnerableLoginApp
{
    static void Main(string[] args)
    {
        CreateDatabase();
        Login();
    }

    private static void CreateDatabase()
    {
        string connectionString = "Data Source=example.db;Version=3;";
        using (var connection = new SQLiteConnection(connectionString))
        {
            connection.Open();

            string createTableQuery = @"
                CREATE TABLE IF NOT EXISTS users (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    username TEXT NOT NULL,
                    password TEXT NOT NULL
                )";
            using (var command = new SQLiteCommand(createTableQuery, connection))
            {
                command.ExecuteNonQuery();
            }

            string insertAdminQuery = @"
                INSERT OR IGNORE INTO users (id, username, password) VALUES (1, 'admin', 'adminpass')";
            using (var command = new SQLiteCommand(insertAdminQuery, connection))
            {
                command.ExecuteNonQuery();
            }

            Console.WriteLine("Database initialized successfully.");
        }
    }

    private static void Login()
    {
        string connectionString = "Data Source=example.db;Version=3;";
        Console.Write("Enter username: ");
        string username = Console.ReadLine();
        Console.Write("Enter password: ");
        string password = Console.ReadLine();
                                                                                                                                           // Vulnerable SQL query concatenating user input directly
        string query = $"SELECT * FROM users WHERE username = '{username}' AND password = '{password}'";

        Console.WriteLine("Executing query: " + query);

        using (var connection = new SQLiteConnection(connectionString))
        {
            connection.Open();
            using (var command = new SQLiteCommand(query, connection))
            {
                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        Console.WriteLine("Login successful");
                    }
                    else
                    {
                        Console.WriteLine("Login failed");
                    }
                }
            }
        }
    }
}























//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
.....................................................................................................................................................................................................

How can this be exploited?

An attacker can craft malicious input to manipulate the query and bypass authentication or access unauthorized data.
Example 1: Bypassing Authentication

If an attacker provides the following input:

    Username: ' OR '1'='1
    Password: anything

The resulting query would look like this:

SELECT * FROM users WHERE username = '' OR '1'='1' AND password = 'anything';

    Breakdown:
        username = '' always evaluates as false.
        '1'='1' always evaluates as true.
        Since the OR condition makes the entire WHERE clause true, the query returns the first user from the users table, logging in the attacker as the admin without knowing the actual password.

Example 2: Accessing Data

An attacker can craft inputs to extract or modify data. For example:

    Username: ' UNION SELECT NULL, sqlite_version(), NULL --
    Password: anything

The query would become:

SELECT * FROM users WHERE username = '' UNION SELECT NULL, sqlite_version(), NULL -- AND password = 'anything';

  

File: Secure Code Review/SQL/Java/sqli.java
import java.sql.*;
import java.util.Scanner;

public class VulnerableLoginApp {

    public static void main(String[] args) {
        createDatabase();
        login();
    }

    private static void createDatabase() {
        String url = "jdbc:sqlite:example.db";
        try (Connection conn = DriverManager.getConnection(url);
             Statement stmt = conn.createStatement()) {

            // Create the users table
            String createTableSQL = "CREATE TABLE IF NOT EXISTS users (id INTEGER PRIMARY KEY, username TEXT, password TEXT)";
            stmt.execute(createTableSQL);

            // Insert a default user (admin)
            String insertUserSQL = "INSERT OR IGNORE INTO users (id, username, password) VALUES (1, 'admin', 'adminpass')";
            stmt.execute(insertUserSQL);

            System.out.println("Database initialized successfully.");
        } catch (SQLException e) {
            System.out.println("Database creation failed: " + e.getMessage());
        }
    }

    private static void login() {
        String url = "jdbc:sqlite:example.db";
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter username: ");
        String username = scanner.nextLine();
        System.out.print("Enter password: ");
        String password = scanner.nextLine();

        // This query concatenates user input directly, making it vulnerable to SQL injection
        String loginQuery = "SELECT * FROM users WHERE username = '" + username + "' AND password = '" + password + "'";

        System.out.println("Executing query: " + loginQuery);

        try (Connection conn = DriverManager.getConnection(url);
             Statement stmt = conn.createStatement()) {

            ResultSet rs = stmt.executeQuery(loginQuery);
            if (rs.next()) {
                System.out.println("Login successful");
            } else {
                System.out.println("Login failed");
            }
        } catch (SQLException e) {
            System.out.println("Database query failed: " + e.getMessage());
        }
    }
}














..............................................................................................................................................................
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

How it could be Exploitation

Suppose the attacker enters the following:

    username: admin' --
    password: (left empty)

The SQL query becomes:

SELECT * FROM users WHERE username = 'admin' --' AND password = ''

    The -- is a SQL comment operator, which causes the rest of the query to be ignored.
    This effectively changes the query to:

    SELECT * FROM users WHERE username = 'admin'

    The attacker is authenticated as the admin user without needing the password.

Further Exploitation Risks

    Data Extraction:
        An attacker could craft inputs to extract sensitive data. For example:
            Input: username = ' OR '1'='1 and password = ' OR '1'='1.
            Query becomes:

        SELECT * FROM users WHERE username = '' OR '1'='1' AND password = '' OR '1'='1'

        This retrieves all rows from the users table.

Database Manipulation:

    An attacker could modify the database using injected SQL commands. For example:
        Input: username = admin'; DROP TABLE users; -- and password = anything.
        Query becomes:

        SELECT * FROM users WHERE username = 'admin'; DROP TABLE users; --' AND password = 'anything'

        This executes two commands: a valid query followed by dropping the users table.

Database Compromise:

    Attackers can potentially execute commands to retrieve all data, delete records, or even escalate access depending on the database configuration.












































